crates/zkevm_circuits/src/bn254/ec_mul/implementation.rs

```diff
@@ -0,0 +1,428 @@
+        let c2 = g1_times_k.to_high();
+
+        // Converting all constants to field elements
+        let mut b1 = convert_uint256_to_field_element(cs, &b1, scalar_field_params);
+        let mut b2 = convert_uint256_to_field_element(cs, &b2, scalar_field_params);
+        let mut c1 = convert_uint256_to_field_element(cs, &c1, scalar_field_params);
+        let mut c2 = convert_uint256_to_field_element(cs, &c2, scalar_field_params);
+        let mut lambda = convert_uint256_to_field_element(cs, &lambda, scalar_field_params);
+
+        // q1 <- c1 * b1
+        // q2 <- c2 * b2
+        let mut q1 = c1.mul(cs, &mut b1);
+        let mut q1 = q1.negated(cs);
+        let mut q2 = c2.mul(cs, &mut b2);
+        let mut q2 = q2.negated(cs);
+
+        // k2 <- q2 - q1
+        let mut k2 = q2.sub(cs, &mut q1);
+        k2.normalize(cs);
+
+        // k2_lambda <- k2 * lambda, k1 <- k - k2_lambda
+        let mut k2_times_lambda = k2.mul(cs, &mut lambda);
+        let mut k1 = scalar.sub(cs, &mut k2_times_lambda);
+        k1.normalize(cs);
+
+        let k1_u256 = convert_field_element_to_uint256(cs, k1.clone());
+        let k2_u256 = convert_field_element_to_uint256(cs, k2.clone());
+
+        let low_pow_2_128 = pow_2_128.to_low();
+
+        // Selecting between k1 and -k1 in Fq
+        let (_, k1_out_of_range) = low_pow_2_128.overflowing_sub(cs, &k1_u256);
+        let k1_negated = k1.negated(cs);
+        let k1 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
+            cs,
+            k1_out_of_range,
+            &k1_negated,
+            &k1,
+        );
+
+        // Selecting between k2 and -k2 in Fq
+        let (_, k2_out_of_range) = low_pow_2_128.overflowing_sub(cs, &k2_u256);
+        let k2_negated = k2.negated(cs);
+        let k2 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
+            cs,
+            k2_out_of_range,
+            &k2_negated,
+            &k2,
+        );
+
+        Self {
+            k1,
+            k2,
+            k1_was_negated: k1_out_of_range,
+            k2_was_negated: k2_out_of_range,
+        }
+    }
+}
+
+pub fn width_4_windowed_multiplication<F, CS>(
+    cs: &mut CS,
+    mut point: BN256SWProjectivePoint<F>,
+    mut scalar: BN256ScalarNNField<F>,
+    base_field_params: &Arc<BN256BaseNNFieldParams>,
+    scalar_field_params: &Arc<BN256ScalarNNFieldParams>,
+) -> BN256SWProjectivePoint<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    // Scalar decomposition
+    scalar.enforce_reduced(cs);
+    let scalar_decomposition = ScalarDecomposition::from(cs, &mut scalar, scalar_field_params);
+
+    // create precomputed table of size 1<<4 - 1
+    // there is no 0 * P in the table, we will handle it below
+    let mut table = Vec::with_capacity(PRECOMPUTATION_TABLE_SIZE);
+    let mut tmp = point.clone();
+    let (mut p_affine, _) = point.convert_to_affine_or_default(cs, BN256Affine::one());
```

Comment:
We have already checked the flag that point is zero, we can reuse those constraint before masking

Reply:
That's right, but we'll only omit doing `NN::is_zero(&mut self.z, cs);` check and that would require making bizarre specific functions and signatures with this extra param, leading to code complications, so I guess that might be not worth it.